From 73aeb0c98b1d5030e735086d6ec782dd7649d345 Mon Sep 17 00:00:00 2001
From: Sanyog Kale <skale@habana.ai>
Date: Sun, 15 May 2022 16:47:57 +0300
Subject: [PATCH] [SW-86865] gaudi3: perform memory initialization

In order to avoid memory exception while copying content in memory,
it is recommended to initialize memory before copying the content.
This patch adds initializing memory before copying image to
destination memory.

Change-Id: I96432bcabcbcf792c99da75378952257686330a1
---

diff --git a/arch/gaudi3/arc/common.c b/arch/gaudi3/arc/common.c
index 57efed0..8ffbded 100644
--- a/arch/gaudi3/arc/common.c
+++ b/arch/gaudi3/arc/common.c
@@ -914,9 +914,27 @@
 	image_entry();
 }
 
+#define MEM_INIT_SIZE_ALIGN				16
 void arcboot_copy_image(uint64_t dst, const void *buf, int len, __unused const char *arch)
 {
 	uint32_t arc_dst = (uint32_t)arcboot_rom_get_arc_addr(dst);
+	int init_len = len;
+
+	/* Skip memory init for ARC1 ICCM and DCCM as its already initialized */
+	if (!((dst >= HL_ARC1_ICCM_FULL_BASE) &&
+	    (dst + len <= HL_ARC1_ICCM_FULL_BASE + ARC1_MEM_SIZE))) {
+
+		/*
+		 * Align length to 16 bytes for memory initialization.
+		 * max length check is already done in fit property verification. If
+		 * image is passed with memory maximum length, then length will
+		 * be already 16 bytes aligned.
+		 */
+		if (len % MEM_INIT_SIZE_ALIGN)
+			init_len += (MEM_INIT_SIZE_ALIGN - (init_len % MEM_INIT_SIZE_ALIGN));
+
+		arcboot_rom_mem_init(dst, init_len);
+	}
 
 	memcpy((void *)arc_dst, buf, len);
 }
